Improve intrinsic sizing and white-space handling around forced line breaks

mrobinson opened the pull request:
This change fixes a few issues:

 - Intrinsic sizing should take into account forced line breaks.
 - Non-preserved white-space should never trigger a line break. This type of white-space can hang off the end of the line and be trimmed later. Hanging white-space is already ignored for the purposes of placing lines among floats.
 - Line breaks should immediately trigger new positioning among floats.

These changes seem a little unrelated, but fixing one of these issues triggers failures due to the other ones. Fixing all of these together leads to test progressions. 

Fixes #30350.

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #30350.
- [x] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
Take forced line breaks into account for intrinsic size

Fixes #30350.

mrobinson commented:
@bors-servo try=wpt-2020

mrobinson commented:
@bors-servo try=wpt-2020

Loirooriol commented on components/layout_2020/flow/inline.rs:
```diff
@@ -707,6 +707,22 @@ impl InlineFormattingContext {
+                                            self.had_non_whitespace_content_yet = true;
+                                            self.current_line.max_content += advance;
+                                            self.forced_line_break();
+                                            self.current_line.max_content = ContentSizes::zero();
```
I get an error:

```
    |
721 | ...                   self.current_line.max_content = ContentSizes::zero();
    |                       -----------------------------   ^^^^^^^^^^^^^^^^^^^^ expected `CSSPixelLength`, found `ContentSizes`
    |                       |
    |                       expected due to the type of this binding
```

mrobinson commented:
@bors-servo try=wpt-2020

mrobinson commented on components/layout_2020/flow/inline.rs:
```diff
@@ -707,6 +707,22 @@ impl InlineFormattingContext {
+                                            .get(run.range.end().to_usize() - 1);
+                                        if last_byte == Some(&b'\n') {
+                                            self.had_non_whitespace_content_yet = true;
+                                            self.current_line.max_content += advance;
```
I think it isn't necessary though we also need to fix the issue regarding soft line break opportunities before forced breaks. I'll open a bug for that issue soon.

mrobinson committed:
Don't linebreak on collapsible whitespace

This whitespace can hang off the end of the line, because it will be
trimmed LineItem layout.

mrobinson commented:
@bors-servo try=wpt-2020

mrobinson commented:
It seems that fixing this issues tickles another issue we have related to trailing non-preserved spacing before `<br>` elements. I'm trying a new approach here where non-preserved spaces *never* trigger line-breaking. These types of spaces are allowed to hang. They will later be trimmed by white-space handling in `LineItem` layout.

Still I'm worried that this work-around might lead to situations where the starting inline margins, borders, and padding will be on a different line than their starting content. That could be a larger issue and perhaps we should solve that when tackling the similar issue we have with ending inline margins, borders, and padding.

mrobinson commented:
I need to look at the failure in 'css/CSS2/css1/c5525-fltmrgn-000.xht' but I'm feeling pretty good about this.

Loirooriol commented:
This is the reduction of the regression:

```html
<div style="font: 15px/1 Ahem; width: 75px; border: solid black; color: magenta">
  <div style="float: left; width: 50px; height: 50px; background: cyan"></div>
  x x
</div>
```

| Bad | Good |
| - | - |
| ![](https://github.com/servo/servo/assets/7477678/a41d4030-a22f-40bd-9c63-0fb8b88dffc1) | ![](https://github.com/servo/servo/assets/7477678/83003507-51c7-4492-a7b2-8051acea5769) |

mrobinson committed:
Update float placement after line breakage

Also don't ever line break for collapsible whitespace.

mrobinson commented:
@bors-servo try=wpt-2020

Loirooriol commented on components/layout_2020/flow/inline.rs:
```diff
@@ -1293,14 +1301,28 @@ impl TextRun {
+            // This prevents whitespace from being added to the beginning of a line. We could
+            // trim it later, but we don't want it to come into play when determining line
+            // width.
+            if is_whitespace && !white_space.preserve_spaces() && !ifc.current_line.has_content {
```
Could reuse `is_non_preserved_whitespace`

mrobinson committed:
Fix a few more test cases and clean up

mrobinson commented:
@bors-servo try=wpt-2020

mrobinson commented:
Okay. I think this is ready for review. Thanks for the feedback!

Loirooriol commented on components/layout_2020/flow/inline.rs:
```diff
@@ -1314,13 +1318,56 @@ impl TextRun {
             }
 
+            let new_advance_from_glyph_run = Length::from(run.glyph_store.total_advance());
+            let new_total_advance = new_advance_from_glyph_run +
```
I wouldn't cal this "advance" if it includes `ifc.current_line.inline_position`. Maybe `new_inline_position`, `inline_end_position` or something.

mrobinson replied:
I've done the following renames:
 - `new_total_advance` -> `new_potential_inline_end_position`
 - `advance_from_text_run` -> `text_run_inline_size`
 - `new_advance_from_glyph_run` -> `advance_from_glyph_run`

mrobinson commented on components/layout_2020/flow/inline.rs:
```diff
@@ -1205,6 +1223,23 @@ impl TextRun {
     }
 
+    fn glyph_run_ends_with_preserved_newline(&self, run: &GlyphRun) -> bool {
+        if !run.glyph_store.is_whitespace() {
+            return false;
+        }
```
I've renamed this to `glyph_run_is_whitespace_ending_with_preserved_newline` to clarify that it doesn't apply to the `TextRun` itself.

mrobinson committed:
Renaming according to review comments

mrobinson commented:
Thanks for the review!
